src/ray/raylet/node_manager.cc

```diff
@@ -2928,6 +2930,79 @@ std::string compact_tag_string(const opencensus::stats::ViewDescriptor &view,
   return result.str();
 }
 
+void NodeManager::HandlePinObjectIDsRequest(const rpc::PinObjectIDsRequest &request,
+                                            rpc::PinObjectIDsReply *reply,
+                                            rpc::SendReplyCallback send_reply_callback) {
+  if (!object_pinning_enabled_) {
+    send_reply_callback(Status::OK(), nullptr, nullptr);
+    return;
+  }
+  WorkerID worker_id = WorkerID::FromBinary(request.owner_address().worker_id());
+  auto it = worker_rpc_clients_.find(worker_id);
+  if (it == worker_rpc_clients_.end()) {
+    auto client = std::unique_ptr<rpc::CoreWorkerClient>(
+        new rpc::CoreWorkerClient(request.owner_address().ip_address(),
+                                  request.owner_address().port(), client_call_manager_));
+    worker_rpc_clients_.emplace(
+        worker_id, std::make_pair<std::unique_ptr<rpc::CoreWorkerClient>, size_t>(
+                       std::move(client), 0));
+    it = worker_rpc_clients_.find(worker_id);
+  }
+  RAY_CHECK(it != worker_rpc_clients_.end());
+
+  // Pin the objects in plasma by getting them and holding a reference to
+  // the returned buffer.
+  std::vector<plasma::ObjectID> plasma_ids;
+  plasma_ids.reserve(request.object_ids_size());
+  for (const auto &object_id_binary : request.object_ids()) {
+    plasma_ids.push_back(plasma::ObjectID::from_binary(object_id_binary));
+  }
+  std::vector<plasma::ObjectBuffer> plasma_results;
+  if (!store_client_.Get(plasma_ids, /*timeout_ms=*/0, &plasma_results).ok()) {
```

Comment:
The `Get` here is a blocking call to plasma, and might can impacts on raylet since it's single threaded. Ideally we should use something that doesn't require a reply, or can wait the reply in a sync way. But we can revisit this later.